### PR TITLE
Fix fold update problem

### PR DESCRIPTION
**Pull request by Shougo**
I have fixed fold update problem.
The fold is updated in insertleave or `nv_replace` if the update is skipped.
It fixes #5299 test failures.

Note: `nv_replace` is normal-mode `r` implementation.
It enters insert mode and change text.
So, the fix is needed...

### CODE REVIEW

**Comment by Shougo** [comment]
Fixed tests.

**Comment by Shougo** [comment]
All tests are passed!

**Comment by justinmk** [comment]
Why not always update folds after leaving insert-mode? That avoids the global flag.

**Comment by Shougo** [comment]
OK.

**Comment by Shougo** [comment]
This fix is OK?

File: src/nvim/fold.c [hunk 1]
```diff
@@ -761,16 +761,12 @@ void clearFolding(win_T *win)
  */
 void foldUpdate(win_T *wp, linenr_T top, linenr_T bot)
 {
-  if (compl_busy || State & INSERT) {
-    return;
-  }
-
-  fold_T      *fp;
-  if (wp->w_buffer->terminal) {
+  if (compl_busy || State & INSERT || wp->w_buffer->terminal) {
```
Comment (justinmk) [inline src/nvim/fold.c]: why not in terminal buffers?

File: src/nvim/edit.c [hunk 2]
```diff
@@ -462,6 +462,8 @@ static void insert_enter(InsertState *s)
     o_lnum = curwin->w_cursor.lnum;
   }
 
+  foldUpdateAll(curwin);
+  foldOpenCursor();
```
Comment (justinmk) [inline src/nvim/edit.c]: What was calling `foldUpdateAll()` before now? We are adding new calls, but can we delete the old one?

File: src/nvim/normal.c [hunk 3]
```diff
@@ -5924,6 +5924,9 @@ static void nv_replace(cmdarg_T *cap)
     curwin->w_set_curswant = true;
     set_last_insert(cap->nchar);
   }
+
+  foldUpdateAll(curwin);
+  foldOpenCursor();
```
Comment (justinmk) [inline src/nvim/normal.c]: Why is this needed? `insert_enter ()` should be enough. The comment above `edit()` mentions that it handles `gr`, `gR`, `R`, etc.

Comment (justinmk) [inline src/nvim/normal.c]: I see now only a special case of `r` is handled by `edit()`.

**Comment by justinmk** [comment]
Merged. Thanks @Shougo !

File: src/nvim/edit.c (hunk 2 above)
Comment (justinmk) [inline src/nvim/edit.c]: @Shougo This seems to open folds on any cursor movement now, even without changes. Why is `foldOpenCursor()` needed? Can we remove it?

**Comment by Shougo** [comment]
Thank you!

**Comment by Shougo** [comment]
I have updated the vim_dev patch.
